File: solutions/go/grains/2/grains.go
package grains

import "errors"

func Square(number int) (r uint64, e error) {
	if number <= 0 || 64 < number {
        e = errors.New("Number out of range")
        return
    }
    r = 1 << (number-1)
    return
}

func Total() uint64 {
	return ^uint64(0)
}

File: solutions/haskell/grains/2/src/Grains.hs
module Grains (square, total) where

import Data.Bits (complement, shift)
import Data.Word (Word64)

square :: Integer -> Maybe Integer
square n
  | 1 <= n && n <= 64 = Just (fromIntegral (1 :: Word64) `shift` (fromInteger n - 1))
  | otherwise = Nothing

total :: Integer
total = (fromIntegral . complement) (0 :: Word64)